File: crates/auv-driver-macos/native/swift/Sources/AuvMacosNative/Ocr.swift
import CoreGraphics
import Foundation
import ImageIO
import Vision

private func emptyOcrTextResponse(request: NativeOcrTextRequest, message: String, recovery: String) -> NativeOcrTextResponse {
  NativeOcrTextResponse(
    recognized_at: nativeNowIso8601(),
    image_path: request.image_path,
    image_width: 0,
    image_height: 0,
    query: request.query,
    exact: request.exact,
    case_sensitive: request.case_sensitive,
    normalized_query: "".intoRustString(),
    crop_enabled: request.crop_enabled,
    crop_x: request.crop_x,
    crop_y: request.crop_y,
    crop_width: request.crop_width,
    crop_height: request.crop_height,
    ocr_scale_factor: 1.0,
    match_indices: RustVec<Int64>(),
    texts: RustVec<RustString>(),
    confidences: RustVec<Double>(),
    x_values: RustVec<Int64>(),
    y_values: RustVec<Int64>(),
    width_values: RustVec<Int64>(),
    height_values: RustVec<Int64>(),
    error_message: message.intoRustString(),
    recovery_hint: recovery.intoRustString()
  )
}

private func emptyVisualRowsResponse(
  request: NativeVisualRowsRequest,
  message: String,
  recovery: String
) -> NativeVisualRowsResponse {
  NativeVisualRowsResponse(
    detected_at: nativeNowIso8601(),
    image_path: request.image_path,
    image_width: 0,
    image_height: 0,
    crop_enabled: request.crop_enabled,
    crop_x: request.crop_x,
    crop_y: request.crop_y,
    crop_width: request.crop_width,
    crop_height: request.crop_height,
    analysis_strip_x: 0,
    analysis_strip_y: 0,
    analysis_strip_width: 0,
    analysis_strip_height: 0,
    row_indices: RustVec<Int64>(),
    x_values: RustVec<Int64>(),
    y_values: RustVec<Int64>(),
    width_values: RustVec<Int64>(),
    height_values: RustVec<Int64>(),
    peak_densities: RustVec<Double>(),
    error_message: message.intoRustString(),
    recovery_hint: recovery.intoRustString()
  )
}

private func loadImage(path: String) -> CGImage? {
  let imageURL = URL(fileURLWithPath: path)
  guard
    let imageSource = CGImageSourceCreateWithURL(imageURL as CFURL, nil),
    let image = CGImageSourceCreateImageAtIndex(imageSource, 0, nil)
  else {
    return nil
  }
  return image
}

private func upscale(_ image: CGImage, factor: CGFloat) -> CGImage? {
  let width = Int((CGFloat(image.width) * factor).rounded())
  let height = Int((CGFloat(image.height) * factor).rounded())
  guard
    let colorSpace = image.colorSpace ?? CGColorSpace(name: CGColorSpace.sRGB),
    let context = CGContext(
      data: nil,
      width: width,
      height: height,
      bitsPerComponent: image.bitsPerComponent,
      bytesPerRow: 0,
      space: colorSpace,
      bitmapInfo: image.bitmapInfo.rawValue
    )
  else {
    return nil
  }

  context.interpolationQuality = .high
  context.draw(image, in: CGRect(x: 0, y: 0, width: width, height: height))
  return context.makeImage()
}

private func foldConfusableScalar(_ scalar: UnicodeScalar) -> UnicodeScalar {
  switch scalar {
  case "|", "!", "l", "I":
    return "i"
  default:
    return scalar
  }
}

private func normalizeForAnchorMatch(_ raw: String, caseSensitive: Bool) -> String {
  let sanitized = nativeSanitize(raw)
  let folded = String(String.UnicodeScalarView(sanitized.unicodeScalars.map(foldConfusableScalar)))
  let lowercased = caseSensitive ? folded : folded.lowercased()
  return String(
    lowercased.unicodeScalars.filter { scalar in
      CharacterSet.alphanumerics.contains(scalar)
    }
  )
}

func find_ocr_text(request: NativeOcrTextRequest) -> NativeOcrTextResponse {
  let imagePath = request.image_path.toString()
  let rawQuery = request.query.toString()
  guard let image = loadImage(path: imagePath) else {
    return emptyOcrTextResponse(
      request: request,
      message: "could not load image for OCR at \(imagePath)",
      recovery: "verify the image artifact path exists and is readable"
    )
  }

  var workingImage = image
  var cropOffsetX: Int64 = 0
  var cropOffsetY: Int64 = 0
  var ocrScaleFactor: CGFloat = 1.0

  if request.crop_enabled {
    let cropRect = CGRect(
      x: Int(request.crop_x),
      y: Int(request.crop_y),
      width: Int(request.crop_width),
      height: Int(request.crop_height)
    ).integral
    guard cropRect.width > 0, cropRect.height > 0 else {
      return emptyOcrTextResponse(
        request: request,
        message: "invalid OCR crop rect \(cropRect)",
        recovery: "pass a positive crop region inside the image bounds"
      )
    }
    guard let croppedImage = image.cropping(to: cropRect) else {
      return emptyOcrTextResponse(
        request: request,
        message: "could not crop OCR image to \(cropRect)",
        recovery: "pass a crop region inside the image bounds"
      )
    }
    cropOffsetX = Int64(cropRect.origin.x.rounded())
    cropOffsetY = Int64(cropRect.origin.y.rounded())
    if let upscaledImage = upscale(croppedImage, factor: 2.0) {
      workingImage = upscaledImage
      ocrScaleFactor = 2.0
    } else {
      workingImage = croppedImage
    }
  }

  let normalizedQuery = request.case_sensitive ? rawQuery : rawQuery.lowercased()
  let normalizedAnchorQuery = normalizeForAnchorMatch(rawQuery, caseSensitive: request.case_sensitive)

  func matches(_ text: String) -> Bool {
    if rawQuery.isEmpty {
      return true
    }
    let normalizedText = request.case_sensitive ? text : text.lowercased()
    let normalizedAnchorText = normalizeForAnchorMatch(text, caseSensitive: request.case_sensitive)
    if request.exact {
      return normalizedText == normalizedQuery || normalizedAnchorText == normalizedAnchorQuery
    }
    return normalizedText.contains(normalizedQuery)
      || normalizedAnchorText.contains(normalizedAnchorQuery)
  }

  let visionRequest = VNRecognizeTextRequest()
  visionRequest.recognitionLevel = .accurate
  visionRequest.usesLanguageCorrection = true
  visionRequest.recognitionLanguages = ["zh-Hans", "zh-Hant", "en-US"]
  if !rawQuery.isEmpty {
    visionRequest.customWords = [rawQuery]
  }
  if #available(macOS 26.0, *) {
    visionRequest.automaticallyDetectsLanguage = true
  }

  let handler = VNImageRequestHandler(cgImage: workingImage, options: [:])
  do {
    try handler.perform([visionRequest])
  } catch {
    return emptyOcrTextResponse(
      request: request,
      message: "vision OCR failed: \(error)",
      recovery: "verify Screen Recording/capture output and retry"
    )
  }

  let observations = (visionRequest.results as? [VNRecognizedTextObservation]) ?? []
  let maxObservations = max(Int(request.max_observations), 1)

  var matchIndices: [Int64] = []
  var texts: [String] = []
  var confidences: [Double] = []
  var xValues: [Int64] = []
  var yValues: [Int64] = []
  var widthValues: [Int64] = []
  var heightValues: [Int64] = []

  for observation in observations.prefix(maxObservations) {
    let candidates = observation.topCandidates(5)
    guard
      let candidate = candidates.first(where: { candidate in
        let text = nativeSanitize(candidate.string)
        return !text.isEmpty && matches(text)
      })
    else { continue }
    let text = nativeSanitize(candidate.string)

    let boundingBox = observation.boundingBox
    let workingX = Int64((boundingBox.minX * CGFloat(workingImage.width)).rounded())
    let workingY = Int64(((1.0 - boundingBox.maxY) * CGFloat(workingImage.height)).rounded())
    let workingWidth = Int64((boundingBox.width * CGFloat(workingImage.width)).rounded())
    let workingHeight = Int64((boundingBox.height * CGFloat(workingImage.height)).rounded())

    let x = cropOffsetX + Int64((CGFloat(workingX) / ocrScaleFactor).rounded())
    let y = cropOffsetY + Int64((CGFloat(workingY) / ocrScaleFactor).rounded())
    let width = Int64((CGFloat(workingWidth) / ocrScaleFactor).rounded())
    let height = Int64((CGFloat(workingHeight) / ocrScaleFactor).rounded())

    matchIndices.append(Int64(matchIndices.count))
    texts.append(text)
    confidences.append(Double(candidate.confidence))
    xValues.append(x)
    yValues.append(y)
    widthValues.append(width)
    heightValues.append(height)
  }

  return NativeOcrTextResponse(
    recognized_at: nativeNowIso8601(),
    image_path: imagePath.intoRustString(),
    image_width: Int64(image.width),
    image_height: Int64(image.height),
    query: nativeSanitize(rawQuery).intoRustString(),
    exact: request.exact,
    case_sensitive: request.case_sensitive,
    normalized_query: normalizedAnchorQuery.intoRustString(),
    crop_enabled: request.crop_enabled,
    crop_x: cropOffsetX,
    crop_y: cropOffsetY,
    crop_width: request.crop_width,
    crop_height: request.crop_height,
    ocr_scale_factor: Double(ocrScaleFactor),
    match_indices: nativeVec(matchIndices),
    texts: nativeStringVec(texts),
    confidences: nativeVec(confidences),
    x_values: nativeVec(xValues),
    y_values: nativeVec(yValues),
    width_values: nativeVec(widthValues),
    height_values: nativeVec(heightValues),
    error_message: nil,
    recovery_hint: nil
  )
}

func find_visual_rows(request: NativeVisualRowsRequest) -> NativeVisualRowsResponse {
  let imagePath = request.image_path.toString()
  guard let image = loadImage(path: imagePath) else {
    return emptyVisualRowsResponse(
      request: request,
      message: "could not load image for visual row detection at \(imagePath)",
      recovery: "verify the image artifact path exists and is readable"
    )
  }

  var workingImage = image
  var cropOffsetX: Int64 = 0
  var cropOffsetY: Int64 = 0

  if request.crop_enabled {
    let cropRect = CGRect(
      x: Int(request.crop_x),
      y: Int(request.crop_y),
      width: Int(request.crop_width),
      height: Int(request.crop_height)
    ).integral
    guard cropRect.width > 0, cropRect.height > 0 else {
      return emptyVisualRowsResponse(
        request: request,
        message: "invalid visual-row crop rect \(cropRect)",
        recovery: "pass a positive crop region inside the image bounds"
      )
    }
    guard let croppedImage = image.cropping(to: cropRect) else {
      return emptyVisualRowsResponse(
        request: request,
        message: "could not crop visual-row image to \(cropRect)",
        recovery: "pass a crop region inside the image bounds"
      )
    }
    workingImage = croppedImage
    cropOffsetX = Int64(cropRect.origin.x.rounded())
    cropOffsetY = Int64(cropRect.origin.y.rounded())
  }

  guard
    let colorSpace = CGColorSpace(name: CGColorSpace.sRGB),
    let context = CGContext(
      data: nil,
      width: workingImage.width,
      height: workingImage.height,
      bitsPerComponent: 8,
      bytesPerRow: 0,
      space: colorSpace,
      bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
    )
  else {
    return emptyVisualRowsResponse(
      request: request,
      message: "could not allocate bitmap context for visual row detection",
      recovery: "retry with a standard PNG screenshot artifact"
    )
  }

  context.draw(workingImage, in: CGRect(x: 0, y: 0, width: workingImage.width, height: workingImage.height))

  guard let rawData = context.data else {
    return emptyVisualRowsResponse(
      request: request,
      message: "visual row bitmap context did not expose pixel data",
      recovery: "retry with a standard PNG screenshot artifact"
    )
  }

  let width = workingImage.width
  let height = workingImage.height
  let bytesPerRow = context.bytesPerRow
  let pixels = rawData.bindMemory(to: UInt8.self, capacity: bytesPerRow * height)

  func pixelComponents(x: Int, y: Int) -> (r: Double, g: Double, b: Double, a: Double) {
    let offset = y * bytesPerRow + x * 4
    return (
      Double(pixels[offset]) / 255.0,
      Double(pixels[offset + 1]) / 255.0,
      Double(pixels[offset + 2]) / 255.0,
      Double(pixels[offset + 3]) / 255.0
    )
  }

  func luminance(_ pixel: (r: Double, g: Double, b: Double, a: Double)) -> Double {
    0.2126 * pixel.r + 0.7152 * pixel.g + 0.0722 * pixel.b
  }

  func saturation(_ pixel: (r: Double, g: Double, b: Double, a: Double)) -> Double {
    let maxValue = max(pixel.r, max(pixel.g, pixel.b))
    let minValue = min(pixel.r, min(pixel.g, pixel.b))
    return maxValue - minValue
  }

  func edgeStrength(x: Int, y: Int) -> Double {
    let pixel = pixelComponents(x: x, y: y)
    let rightPixel = pixelComponents(x: min(width - 1, x + 1), y: y)
    let downPixel = pixelComponents(x: x, y: min(height - 1, y + 1))
    return
      abs(luminance(pixel) - luminance(rightPixel))
      + abs(luminance(pixel) - luminance(downPixel))
  }

  func isActive(x: Int, y: Int) -> Bool {
    let pixel = pixelComponents(x: x, y: y)
    guard pixel.a > 0.05 else { return false }
    return edgeStrength(x: x, y: y) >= 0.10 || saturation(pixel) > 0.24
  }

  let stripLeft = max(0, Int((Double(width) * 0.02).rounded()))
  let stripRight = min(width, max(stripLeft + 1, Int((Double(width) * 0.24).rounded())))
  let stripWidth = max(1, stripRight - stripLeft)

  var rowSignal = Array(repeating: 0.0, count: height)
  for y in 0..<height {
    var activeCount = 0
    for x in stripLeft..<stripRight {
      if isActive(x: x, y: y) {
        activeCount += 1
      }
    }
    rowSignal[y] = Double(activeCount) / Double(stripWidth)
  }

  let smoothingRadius = 2
  var smoothedSignal = Array(repeating: 0.0, count: height)
  for y in 0..<height {
    let lower = max(0, y - smoothingRadius)
    let upper = min(height - 1, y + smoothingRadius)
    let window = rowSignal[lower...upper]
    smoothedSignal[y] = window.reduce(0.0, +) / Double(window.count)
  }

  let rowThreshold = 0.018
  let maxGap = 10
  let minBandHeight = 28
  let maxBandHeight = 220

  var rawBands = [(start: Int, end: Int)]()
  var bandStart: Int? = nil
  var gap = 0
  for y in 0..<height {
    if smoothedSignal[y] >= rowThreshold {
      if bandStart == nil {
        bandStart = y
      }
      gap = 0
    } else if let start = bandStart {
      gap += 1
      if gap > maxGap {
        rawBands.append((start: start, end: y - gap))
        bandStart = nil
        gap = 0
      }
    }
  }
  if let start = bandStart {
    rawBands.append((start: start, end: height - 1))
  }

  let filteredBands = rawBands.filter { band in
    let bandHeight = band.end - band.start + 1
    return bandHeight >= minBandHeight && bandHeight <= maxBandHeight
  }

  var rowIndices: [Int64] = []
  var xValues: [Int64] = []
  var yValues: [Int64] = []
  var widthValues: [Int64] = []
  var heightValues: [Int64] = []
  var peakDensities: [Double] = []

  for (bandIndex, band) in filteredBands.enumerated() {
    let bandTop = max(0, band.start - 6)
    let bandBottom = min(height - 1, band.end + 6)
    let bandHeight = bandBottom - bandTop + 1

    var leftX: Int?
    var rightX: Int?
    for x in 0..<width {
      var activeCount = 0
      for y in bandTop...bandBottom {
        if isActive(x: x, y: y) {
          activeCount += 1
        }
      }
      let columnDensity = Double(activeCount) / Double(max(1, bandHeight))
      if columnDensity >= 0.04 {
        if leftX == nil {
          leftX = x
        }
        rightX = x
      }
    }

    let visualLeft = max(0, (leftX ?? stripLeft) - 8)
    let visualRight = min(width - 1, (rightX ?? (width - 1)) + 8)
    let visualWidth = max(1, visualRight - visualLeft + 1)
    let peakDensity = smoothedSignal[band.start...band.end].max() ?? 0.0

    rowIndices.append(Int64(bandIndex))
    xValues.append(cropOffsetX + Int64(visualLeft))
    yValues.append(cropOffsetY + Int64(bandTop))
    widthValues.append(Int64(visualWidth))
    heightValues.append(Int64(bandHeight))
    peakDensities.append(peakDensity)
  }

  return NativeVisualRowsResponse(
    detected_at: nativeNowIso8601(),
    image_path: imagePath.intoRustString(),
    image_width: Int64(image.width),
    image_height: Int64(image.height),
    crop_enabled: request.crop_enabled,
    crop_x: cropOffsetX,
    crop_y: cropOffsetY,
    crop_width: request.crop_width,
    crop_height: request.crop_height,
    analysis_strip_x: Int64(stripLeft),
    analysis_strip_y: 0,
    analysis_strip_width: Int64(stripWidth),
    analysis_strip_height: Int64(height),
    row_indices: nativeVec(rowIndices),
    x_values: nativeVec(xValues),
    y_values: nativeVec(yValues),
    width_values: nativeVec(widthValues),
    height_values: nativeVec(heightValues),
    peak_densities: nativeVec(peakDensities),
    error_message: nil,
    recovery_hint: nil
  )
}

File: crates/auv-driver-macos/src/native/ocr.rs
// File: src/driver/macos/native/ocr.rs
use std::path::{Path, PathBuf};

#[cfg(target_os = "macos")]
use super::binding::ffi::{
  NativeOcrRgbaRequest, NativeOcrTextRequest, NativeOcrTextResponse, NativeVisualRowsRequest, NativeVisualRowsResponse, find_ocr_text,
  find_ocr_text_rgba, find_visual_rows,
};
use super::types::{AuvResult, DetectedScreenRows, ObservedOcrRow, ObservedRect, OcrTextMatch, OcrTextSnapshot};

#[derive(Clone, Debug, PartialEq)]
pub struct NativeOcrTextCapture {
  pub snapshot: OcrTextSnapshot,
  pub normalized_query: String,
  pub crop_rect: Option<ObservedRect>,
  pub ocr_scale_factor: f64,
}

#[derive(Clone, Debug, PartialEq)]
pub struct NativeVisualRowsCapture {
  pub rows: DetectedScreenRows,
  pub detected_at: String,
  pub image_path: PathBuf,
  pub image_width: i64,
  pub image_height: i64,
  pub crop_rect: Option<ObservedRect>,
  pub analysis_strip: ObservedRect,
  pub peak_densities: Vec<f64>,
}

#[cfg(target_os = "macos")]
#[allow(clippy::too_many_arguments)]
pub fn find_text(
  image_path: &Path,
  query: &str,
  exact: bool,
  case_sensitive: bool,
  max_observations: i64,
  custom_words: &[String],
  recognition_languages: Option<&[String]>,
  crop_region: Option<&ObservedRect>,
) -> AuvResult<NativeOcrTextCapture> {
  let crop = crop_region.cloned().unwrap_or(ObservedRect {
    x: 0,
    y: 0,
    width: 0,
    height: 0,
  });
  decode_ocr_text_response(DecodedOcrTextResponse::from(find_ocr_text(NativeOcrTextRequest {
    image_path: image_path.display().to_string(),
    query: query.to_string(),
    exact,
    case_sensitive,
    max_observations,
    custom_words: custom_words.to_vec(),
    recognition_languages: recognition_languages.map(<[String]>::to_vec).unwrap_or_default(),
    crop_enabled: crop_region.is_some(),
    crop_x: crop.x,
    crop_y: crop.y,
    crop_width: crop.width,
    crop_height: crop.height,
  })))
}

#[cfg(target_os = "macos")]
#[allow(clippy::too_many_arguments)]
pub fn find_text_in_rgba(
  rgba_bytes: Vec<u8>,
  image_width: i64,
  image_height: i64,
  query: &str,
  exact: bool,
  case_sensitive: bool,
  max_observations: i64,
  custom_words: &[String],
  recognition_languages: Option<&[String]>,
  crop_region: Option<&ObservedRect>,
) -> AuvResult<NativeOcrTextCapture> {
  let crop = crop_region.cloned().unwrap_or(ObservedRect {
    x: 0,
    y: 0,
    width: 0,
    height: 0,
  });
  decode_ocr_text_response(DecodedOcrTextResponse::from(find_ocr_text_rgba(NativeOcrRgbaRequest {
    image_width,
    image_height,
    rgba_bytes,
    query: query.to_string(),
    exact,
    case_sensitive,
    max_observations,
    custom_words: custom_words.to_vec(),
    recognition_languages: recognition_languages.map(<[String]>::to_vec).unwrap_or_default(),
    crop_enabled: crop_region.is_some(),
    crop_x: crop.x,
    crop_y: crop.y,
    crop_width: crop.width,
    crop_height: crop.height,
  })))
}

#[cfg(not(target_os = "macos"))]
#[allow(clippy::too_many_arguments)]
pub fn find_text_in_rgba(
  _rgba_bytes: Vec<u8>,
  _image_width: i64,
  _image_height: i64,
  _query: &str,
  _exact: bool,
  _case_sensitive: bool,
  _max_observations: i64,
  _custom_words: &[String],
  _recognition_languages: Option<&[String]>,
  _crop_region: Option<&ObservedRect>,
) -> AuvResult<NativeOcrTextCapture> {
  Err("macOS native OCR text detection is unsupported on this target".to_string())
}

#[cfg(not(target_os = "macos"))]
pub fn find_text(
  _image_path: &Path,
  _query: &str,
  _exact: bool,
  _case_sensitive: bool,
  _max_observations: i64,
  _custom_words: &[String],
  _recognition_languages: Option<&[String]>,
  _crop_region: Option<&ObservedRect>,
) -> AuvResult<NativeOcrTextCapture> {
  Err("macOS native OCR text detection is unsupported on this target".to_string())
}

#[cfg(target_os = "macos")]
pub fn find_rows(image_path: &Path, crop_region: Option<&ObservedRect>) -> AuvResult<NativeVisualRowsCapture> {
  let crop = crop_region.cloned().unwrap_or(ObservedRect {
    x: 0,
    y: 0,
    width: 0,
    height: 0,
  });
  decode_visual_rows_response(DecodedVisualRowsResponse::from(find_visual_rows(NativeVisualRowsRequest {
    image_path: image_path.display().to_string(),
    crop_enabled: crop_region.is_some(),
    crop_x: crop.x,
    crop_y: crop.y,
    crop_width: crop.width,
    crop_height: crop.height,
  })))
}

#[cfg(not(target_os = "macos"))]
pub fn find_rows(_image_path: &Path, _crop_region: Option<&ObservedRect>) -> AuvResult<NativeVisualRowsCapture> {
  Err("macOS native visual row detection is unsupported on this target".to_string())
}

pub fn decode_ocr_text_response(response: DecodedOcrTextResponse) -> AuvResult<NativeOcrTextCapture> {
  if response.error_message.is_some() {
    return super::error::native_result("find_ocr_text", None, response.error_message, response.recovery_hint);
  }

  let count = response.match_indices.len();
  let lengths = [
    response.texts.len(),
    response.confidences.len(),
    response.x_values.len(),
    response.y_values.len(),
    response.width_values.len(),
    response.height_values.len(),
  ];
  if lengths.iter().any(|length| *length != count) {
    return Err("native OCR text response had mismatched OCR match vector lengths".to_string());
  }

  let matches = (0..count)
    .map(|index| {
      let match_index = usize::try_from(response.match_indices[index])
        .map_err(|error| format!("native OCR text response had invalid match index {}: {error}", response.match_indices[index]))?;
      Ok(OcrTextMatch {
        match_index,
        text: response.texts[index].clone(),
        confidence: response.confidences[index],
        bounds: ObservedRect {
          x: response.x_values[index],
          y: response.y_values[index],
          width: response.width_values[index],
          height: response.height_values[index],
        },
      })
    })
    .collect::<AuvResult<Vec<_>>>()?;

  let crop_rect = response.crop_enabled.then_some(ObservedRect {
    x: response.crop_x,
    y: response.crop_y,
    width: response.crop_width,
    height: response.crop_height,
  });

  Ok(NativeOcrTextCapture {
    snapshot: OcrTextSnapshot {
      recognized_at: response.recognized_at,
      image_path: PathBuf::from(response.image_path),
      image_width: response.image_width,
      image_height: response.image_height,
      query: response.query,
      exact: response.exact,
      case_sensitive: response.case_sensitive,
      matches,
    },
    normalized_query: response.normalized_query,
    crop_rect,
    ocr_scale_factor: response.ocr_scale_factor,
  })
}

pub fn decode_visual_rows_response(response: DecodedVisualRowsResponse) -> AuvResult<NativeVisualRowsCapture> {
  if response.error_message.is_some() {
    return super::error::native_result("find_visual_rows", None, response.error_message, response.recovery_hint);
  }

  let count = response.row_indices.len();
  let lengths = [
    response.x_values.len(),
    response.y_values.len(),
    response.width_values.len(),
    response.height_values.len(),
    response.peak_densities.len(),
  ];
  if lengths.iter().any(|length| *length != count) {
    return Err("native visual rows response had mismatched row vector lengths".to_string());
  }

  let rows = (0..count)
    .map(|index| {
      let row_index = usize::try_from(response.row_indices[index])
        .map_err(|error| format!("native visual rows response had invalid row index {}: {error}", response.row_indices[index]))?;
      Ok(ObservedOcrRow {
        row_index,
        source: "visual-bands".to_string(),
        bounds: ObservedRect {
          x: response.x_values[index],
          y: response.y_values[index],
          width: response.width_values[index],
          height: response.height_values[index],
        },
        text_fragments: vec![],
      })
    })
    .collect::<AuvResult<Vec<_>>>()?;

  let crop_rect = response.crop_enabled.then_some(ObservedRect {
    x: response.crop_x,
    y: response.crop_y,
    width: response.crop_width,
    height: response.crop_height,
  });
  let analysis_strip = ObservedRect {
    x: response.analysis_strip_x,
    y: response.analysis_strip_y,
    width: response.analysis_strip_width,
    height: response.analysis_strip_height,
  };
  Ok(NativeVisualRowsCapture {
    rows: DetectedScreenRows {
      strategy: "visual-bands".to_string(),
      raw_match_count: 0,
      filtered_match_count: 0,
      rows,
    },
    detected_at: response.detected_at,
    image_path: PathBuf::from(response.image_path),
    image_width: response.image_width,
    image_height: response.image_height,
    crop_rect,
    analysis_strip,
    peak_densities: response.peak_densities,
  })
}

#[derive(Clone, Debug)]
pub struct DecodedOcrTextResponse {
  pub recognized_at: String,
  pub image_path: String,
  pub image_width: i64,
  pub image_height: i64,
  pub query: String,
  pub exact: bool,
  pub case_sensitive: bool,
  pub normalized_query: String,
  pub crop_enabled: bool,
  pub crop_x: i64,
  pub crop_y: i64,
  pub crop_width: i64,
  pub crop_height: i64,
  pub ocr_scale_factor: f64,
  pub match_indices: Vec<i64>,
  pub texts: Vec<String>,
  pub confidences: Vec<f64>,
  pub x_values: Vec<i64>,
  pub y_values: Vec<i64>,
  pub width_values: Vec<i64>,
  pub height_values: Vec<i64>,
  pub error_message: Option<String>,
  pub recovery_hint: Option<String>,
}

#[derive(Clone, Debug)]
pub struct DecodedVisualRowsResponse {
  pub detected_at: String,
  pub image_path: String,
  pub image_width: i64,
  pub image_height: i64,
  pub crop_enabled: bool,
  pub crop_x: i64,
  pub crop_y: i64,
  pub crop_width: i64,
  pub crop_height: i64,
  pub analysis_strip_x: i64,
  pub analysis_strip_y: i64,
  pub analysis_strip_width: i64,
  pub analysis_strip_height: i64,
  pub row_indices: Vec<i64>,
  pub x_values: Vec<i64>,
  pub y_values: Vec<i64>,
  pub width_values: Vec<i64>,
  pub height_values: Vec<i64>,
  pub peak_densities: Vec<f64>,
  pub error_message: Option<String>,
  pub recovery_hint: Option<String>,
}

#[cfg(target_os = "macos")]
impl From<NativeOcrTextResponse> for DecodedOcrTextResponse {
  fn from(value: NativeOcrTextResponse) -> Self {
    Self {
      recognized_at: value.recognized_at,
      image_path: value.image_path,
      image_width: value.image_width,
      image_height: value.image_height,
      query: value.query,
      exact: value.exact,
      case_sensitive: value.case_sensitive,
      normalized_query: value.normalized_query,
      crop_enabled: value.crop_enabled,
      crop_x: value.crop_x,
      crop_y: value.crop_y,
      crop_width: value.crop_width,
      crop_height: value.crop_height,
      ocr_scale_factor: value.ocr_scale_factor,
      match_indices: value.match_indices,
      texts: value.texts,
      confidences: value.confidences,
      x_values: value.x_values,
      y_values: value.y_values,
      width_values: value.width_values,
      height_values: value.height_values,
      error_message: value.error_message,
      recovery_hint: value.recovery_hint,
    }
  }
}

#[cfg(target_os = "macos")]
impl From<NativeVisualRowsResponse> for DecodedVisualRowsResponse {
  fn from(value: NativeVisualRowsResponse) -> Self {
    Self {
      detected_at: value.detected_at,
      image_path: value.image_path,
      image_width: value.image_width,
      image_height: value.image_height,
      crop_enabled: value.crop_enabled,
      crop_x: value.crop_x,
      crop_y: value.crop_y,
      crop_width: value.crop_width,
      crop_height: value.crop_height,
      analysis_strip_x: value.analysis_strip_x,
      analysis_strip_y: value.analysis_strip_y,
      analysis_strip_width: value.analysis_strip_width,
      analysis_strip_height: value.analysis_strip_height,
      row_indices: value.row_indices,
      x_values: value.x_values,
      y_values: value.y_values,
      width_values: value.width_values,
      height_values: value.height_values,
      peak_densities: value.peak_densities,
      error_message: value.error_message,
      recovery_hint: value.recovery_hint,
    }
  }
}

#[cfg(test)]
#[path = "ocr_test.rs"]
mod tests;
